pkg/gadget-context/run.go

```diff
@@ -88,79 +90,115 @@ func (c *GadgetContext) initAndPrepareOperators(paramValues api.ParamValues) ([]
+			}
 			return fmt.Errorf("starting operator %q: %w", opInst.Name(), err)
 		}
+		started = append(started, opInst)
 	}
 	return nil
 }
 
-func (c *GadgetContext) stop(dataOperatorInstances []operators.DataOperatorInstance) {
-	// Stop/DeInit in reverse order
-	for i := len(dataOperatorInstances) - 1; i >= 0; i-- {
-		opInst := dataOperatorInstances[i]
+func (c *GadgetContext) stop() error {
+	// Stop in reverse order
+	for i := len(c.dataOperatorInstances) - 1; i >= 0; i-- {
+		opInst := c.dataOperatorInstances[i]
 		c.Logger().Debugf("stopping op %q", opInst.Name())
 		err := opInst.Stop(c)
 		if err != nil {
-			c.Logger().Errorf("stopping operator %q: %v", opInst.Name(), err)
+			return fmt.Errorf("stopping operator %q: %w", opInst.Name(), err)
 		}
 	}
-	// Stop/DeInit in reverse order
-	for i := len(dataOperatorInstances) - 1; i >= 0; i-- {
-		opInst := dataOperatorInstances[i]
-		postStop, ok := opInst.(operators.PostStop)
-		if !ok {
-			continue
+	return nil
+}
+
+func (c *GadgetContext) postStop() error {
+	// PostStop in reverse order
+	for i := len(c.dataOperatorInstances) - 1; i >= 0; i-- {
+		opInst := c.dataOperatorInstances[i]
+		if postStop, ok := opInst.(operators.PostStop); ok {
+			c.Logger().Debugf("post-stopping op %q", opInst.Name())
+			err := postStop.PostStop(c)
+			if err != nil {
+				return fmt.Errorf("post-stopping operator %q: %w", opInst.Name(), err)
+			}
 		}
-		c.Logger().Debugf("post-stopping op %q", opInst.Name())
-		err := postStop.PostStop(c)
+	}
+	return nil
+}
+
+func (c *GadgetContext) close() error {
+	var result error
+
+	// Close in reverse order
+	for i := len(c.dataOperatorInstances) - 1; i >= 0; i-- {
+		opInst := c.dataOperatorInstances[i]
+		c.Logger().Debugf("closing op %q", opInst.Name())
+		err := opInst.Close(c)
 		if err != nil {
-			c.Logger().Errorf("post-stopping operator %q: %v", opInst.Name(), err)
+			result = multierror.Append(result, fmt.Errorf("closing operator %q: %w", opInst.Name(), err))
 		}
 	}
+	return result
 }
 
 func (c *GadgetContext) PrepareGadgetInfo(paramValues api.ParamValues) error {
-	_, err := c.initAndPrepareOperators(paramValues)
+	err := c.initAndPrepareOperators(paramValues)
+	c.close()
 	return err
 }
 
 func (c *GadgetContext) Run(paramValues api.ParamValues) error {
 	defer c.cancel()
+	defer c.close()
 
-	dataOperatorInstances, err := c.initAndPrepareOperators(paramValues)
+	err := c.initAndPrepareOperators(paramValues)
 	if err != nil {
 		return fmt.Errorf("initializing and preparing operators: %w", err)
 	}
 
-	if err := c.start(dataOperatorInstances); err != nil {
+	if err := c.preStart(); err != nil {
+		return fmt.Errorf("pre-starting operators: %w", err)
+	}
+
+	if err := c.start(); err != nil {
 		return fmt.Errorf("starting operators: %w", err)
 	}
 
 	c.Logger().Debugf("running...")
 
 	WaitForTimeoutOrDone(c)
-	c.stop(dataOperatorInstances)
+	if err := c.stop(); err != nil {
+		return fmt.Errorf("stopping operators: %w", err)
```

Comment:
I'm finding all of this a little bit complicated as the logic to handle a failure and "reverting the chain" is not that easy to implement. What about a simplified model:
- Hooks are called in the following order: Init() -> PreStart() -> Stop() -> PostStop( ) -> Close()
- If any of the hooks fails, the next ones are skipped 
- Close() is always called
- Some hooks can be skipped even if the previous ones worked fine (in case another operator failed)